### DISCUSSION TITLE
SQLAlchemy 2.0 Core Type detection for whole table selects not working with Mypy

### DISCUSSION THREAD

**Discussion by igorbga** [sqlalchemy/sqlalchemy#9801] [Usage Questions] [answered]
SQLAlchemy version is 2.0.13
Mypy version is 1.3.0
Mypy extensions version 1.0.0

I'm using VS Code.
I'm having some trouble wile trying Mypy to correctly detect the types of my rows.
I have simplified the issue to the following code.

```python
from sqlalchemy import Column, String, Table, create_engine, select, MetaData


engine = create_engine("whatever")
metadata_obj = MetaData()

text_col = Column("text", String, nullable=False)
test_table = Table("test_table", metadata_obj, text_col)

with engine.connect() as connection:
    stmt = select(text_col)
    stmt2 = select(test_table)
    stmt3 = select(test_table.c.text)
```

If I hover on stmt it is said to be typed "Select[Tuple[str]]", as expected.

But if I hover on stmt2 it is types "Select[Any]"
For stmt3 it says "Select[Tuple[Any]]".

Does this mean that for typing to work "completely" when using SQLAlchemy Core selects I have to keep a variable pointing to each and all the columns and add them one by one to the select? 

Thanks

**Comment by zzzeek** [comment, marked as answer]
hi -

Table is a dynamically built construct, so Mypy has no way to know that it has an attribute named ".c.text" nor does it have any way to know that it contains a String column.   This is why all the mypy stuff in the docs emphasizes that this works better with ORM declarative classes, *even* if you are only using Core:


```py
from sqlalchemy import Column, String, Table, create_engine, select, MetaData

from sqlalchemy.orm import Mapped

engine = create_engine("whatever")
metadata_obj = MetaData()

text_col = Column("text", String, nullable=False)
test_table = Table("test_table", metadata_obj, text_col)

class TestTable:
    __table__ = test_table

    text: Mapped[str]

with engine.connect() as connection:
    stmt = select(text_col)
    stmt2 = select(TestTable)
    stmt3 = select(TestTable.text)
```

Mypy can only know that "obja.objb" has a particular type if "obja" is a class structure and "objb" is an attribute of that class; *or* if obja is a typing.NamedTuple.   There is **no** other way in Python typing right now for the type of `obja.objb` to be known by typing tools without custom plugins.

**Reply by zzzeek** [reply]
Here's the best I can do using only Table:

```py
from typing import NamedTuple

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import Table

engine = create_engine("sqlite://")
metadata_obj = MetaData()

text_col = Column("text", String, nullable=False)

class TestTable(Table):
    class cols(NamedTuple):
        text: Column[str]

    c: cols

test_table: TestTable = TestTable("test_table", metadata_obj, text_col)

with engine.connect() as connection:
    stmt = select(text_col)
    stmt2 = select(test_table.c)
    stmt3 = select(test_table.c.text)
```

**Reply by igorbga** [reply]
Ok, I had read 1.4 documentation some months ago and now had just grasped around the migration to 2.0 and the details about using ORM constructs in CORE only usages was not obvius for me.

In fact, your last suggestion, although interesting an certainly maybe a faster option for not having to change all the accesses to the rows, has itself some typing issues due to the fact that Table already defines c as being of type `ReadOnlyColumnCollection[str, Column[Any]]`.

The nearest way to mixing both styles I've been able to achieve is this:

```python

from sqlalchemy import create_engine, select
from sqlalchemy.orm import DeclarativeBase, Mapped


engine = create_engine("whatever")


class Base(DeclarativeBase):
    ...


class TestTable(Base):
    __tablename__ = "test_table"

    text: Mapped[str]


with engine.connect() as connection:
    stmt = select(TestTable)

    result = connection.execute(stmt)
    row = result.first()
    if row is None:
        raise Exception
    value = row.t[0]
    value.text

    result2 = connection.scalars(stmt)
    row2 = result2.first()
    if row2 is None:
        raise Exception
    value2 = row2.text

```

But that obviously implies changing the way I access and manage the rows and in fact, to be honest, this is more of a "ORM like usage without UnitOfWork" than "Core usage with ORM Declarative table definition".

There is probably more twisting that can be done to solve the `ReadOnlyColumnCollection` but my application is not still in production and although not trivial is not too big either so i think i will be able to switch it to ORM completeley in a couple of days. 
Besides, I was already missing the functionality provided by a UnitOfWork in some points.

Thanks @zzzeek for your aclarations.

**Comment by sk-** [comment]
@zzzeek would it be possible for the mypy plugin to handle these cases?

**Reply by zzzeek** [reply]
hi -

the mypy plugin is dead.  please upgrade to 2.0 and follow the steps at https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping

**Reply by CaselIT** [reply]
To expand in 2.0 it's deprecated, while it's removed from 2.1 completely

**Reply by sk-** [reply]
Thanks for the quick reply. We were already using version 2.0, but I had forgotten the plugin was deprecated and no longer in use. Sorry for the noise. 

Are there any plans to provide better typing support for core tables?

**Reply by zzzeek** [reply]
my selected answer here indicates why it's not possible with the current capabilities of pep-484

**Reply by CaselIT** [reply]
There is kind of of a poc here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5456, but I 'm.not sure it will be done in 2.1
